cellxgene_resources/cellxgene_mods.py

```diff
@@ -636,3 +636,107 @@ def compare_revision(collection):
+    return (fem,male)
+
+
+def calculate_sex(fm_dict):
+    """
+    Input: output dictionary from generate_fm_dict()
+    Output: dataframe of both male and female donors that have 100+ total counts and stats
+    """
+    try:
+        female_df = fm_dict['female'][2]
+        female_df['female_sum'] = female_df.sum(numeric_only=True, axis=1)
+        male_df = fm_dict['male'][2]
+        male_df['male_sum'] = male_df.sum(numeric_only=True, axis=1)
+        male_female_df = pd.merge(male_df, female_df, how='left', on='donor_id')
+        male_female_df['total_sum'] = male_female_df[['female_sum','male_sum']].sum(numeric_only=True, axis=1)
+        donors_to_remove = male_female_df[male_female_df.total_sum < 100].donor_id.unique()  # Remove donors that have less than 100 total counts
+        if len(donors_to_remove) > 0:
+            print('Donors with < 100 total counts dropped: ', *(donors_to_remove), sep='\n')
+            male_female_df.drop(male_female_df[male_female_df.total_sum < 100].index, inplace=True)
+        #Calculate ratio and assign sex
+        male_female_df['male_to_female'] = male_female_df['male_sum']/male_female_df['female_sum']
+        male_female_df['scRNAseq_sex'] = male_female_df.apply(lambda x: assign_sex(x['male_to_female']), axis=1)
+        return male_female_df
+    except Exception as e:
+        print(e)
+
+
+def evaluate_donors_sex(adata):
+    genes_file = 'ref_files/sex_analysis_genes.json'
+    genes = json.load(open(genes_file))
+    female_ids = genes['female']
+    male_ids = genes['male']
+
+    fields_to_include = ['colllection_id','dataset_id','donor_id', 'sex_ontology_term_id']
+    metadata_list = [f for f in fields_to_include if f in adata.obs_keys()]
+
+    smart_assay_list = ['EFO:0010184','EFO:0008931','EFO:0008930','EFO:0010022','EFO:0700016','EFO:0022488','EFO:0008442']
+    obs_to_keep = adata.obs.index[adata.obs['assay_ontology_term_id'].isin(smart_assay_list) == False]
+    adata_new = adata[obs_to_keep, : ]
+    print(f'Smart-seq removal: {adata.shape[0] - adata_new.shape[0]} obs removed.')
+
+    if adata_new.raw:
+        female_adata = adata_new.raw[:,adata_new.raw.var.index.isin(female_ids)]
+        male_adata = adata_new.raw[:,adata_new.raw.var.index.isin(male_ids)]
+    else:
+        female_adata = adata_new[:,adata_new.var.index.isin(female_ids)]
+        male_adata = adata_new[:,adata_new.var.index.isin(male_ids)]
+
+    genes_found = check_percent(female_adata,male_adata,female_ids,male_ids)
+
+    if 0 in genes_found:
+        print('Cannot calculate sex - male/female/both gene set(s) not found in dataset.')
+    else:
+        fm_counts_dict = generate_fm_dict(female_ids,female_adata,male_ids,male_adata,adata_new)
+        donor_sex_df = calculate_sex(fm_counts_dict)
+        donor_sex_df = donor_sex_df[['donor_id','male_sum','female_sum','total_sum','male_to_female','scRNAseq_sex']].merge(adata_new.obs[metadata_list].drop_duplicates(), on='donor_id', how='left')
```

Comment:
very long line. if we do keep all of the *_sum fields, I would break this into two - 1 to set the columns & a 2nd to do the merge